File: go/retries.go
package modal

import (
	"fmt"
	"time"
)

// RetriesParams are options for creating a Retries policy.
type RetriesParams struct {
	BackoffCoefficient *float32       // Multiplier for exponential backoff. Defaults to 2.0.
	InitialDelay       *time.Duration // Defaults to 1s.
	MaxDelay           *time.Duration // Defaults to 60s.
}

// Retries represents retry policy configuration for a Modal Function/Cls.
type Retries struct {
	MaxRetries         int
	BackoffCoefficient float32
	InitialDelay       time.Duration
	MaxDelay           time.Duration
}

// NewRetries creates a new Retries configuration.
func NewRetries(maxRetries int, params *RetriesParams) (*Retries, error) {
	backoffCoefficient := float32(2.0)
	initialDelay := 1 * time.Second
	maxDelay := 60 * time.Second

	if params != nil {
		if params.BackoffCoefficient != nil {
			backoffCoefficient = *params.BackoffCoefficient
		}
		if params.InitialDelay != nil {
			initialDelay = *params.InitialDelay
		}
		if params.MaxDelay != nil {
			maxDelay = *params.MaxDelay
		}
	}

	r := &Retries{
		MaxRetries:         maxRetries,
		BackoffCoefficient: backoffCoefficient,
		InitialDelay:       initialDelay,
		MaxDelay:           maxDelay,
	}

	if r.MaxRetries < 0 || r.MaxRetries > 10 {
		return nil, fmt.Errorf("invalid maxRetries: %d. Must be between 0 and 10", r.MaxRetries)
	}

	if r.BackoffCoefficient < 1.0 || r.BackoffCoefficient > 10.0 {
		return nil, fmt.Errorf("invalid backoffCoefficient: %f. Must be between 1.0 and 10.0", r.BackoffCoefficient)
	}

	if r.InitialDelay < 0 || r.InitialDelay > 60*time.Second {
		return nil, fmt.Errorf("invalid initialDelay: %v. Must be between 0 and 60s", r.InitialDelay)
	}

	if r.MaxDelay < 1*time.Second || r.MaxDelay > 60*time.Second {
		return nil, fmt.Errorf("invalid maxDelay: %v. Must be between 1s and 60s", r.MaxDelay)
	}

	return r, nil
}

File: js/src/retries.ts
/** Retry policy configuration for a Modal Function/Cls. */
export class Retries {
  readonly maxRetries: number;
  readonly backoffCoefficient: number;
  readonly initialDelayMs: number;
  readonly maxDelayMs: number;

  constructor(params: {
    maxRetries: number;
    backoffCoefficient?: number;
    initialDelayMs?: number;
    maxDelayMs?: number;
  }) {
    const {
      maxRetries,
      backoffCoefficient = 2.0,
      initialDelayMs = 1000,
      maxDelayMs = 60_000,
    } = params;

    if (maxRetries < 0 || maxRetries > 10) {
      throw new Error(
        `Invalid maxRetries: ${maxRetries}. Must be between 0 and 10.`,
      );
    }

    if (backoffCoefficient < 1.0 || backoffCoefficient > 10.0) {
      throw new Error(
        `Invalid backoffCoefficient: ${backoffCoefficient}. Must be between 1.0 and 10.0`,
      );
    }

    if (initialDelayMs < 0 || initialDelayMs > 60_000) {
      throw new Error(
        `Invalid initialDelayMs: ${initialDelayMs}. Must be between 0 and 60000 ms.`,
      );
    }

    if (maxDelayMs < 1_000 || maxDelayMs > 60_000) {
      throw new Error(
        `Invalid maxDelayMs: ${maxDelayMs}. Must be between 1000 and 60000 ms.`,
      );
    }

    this.maxRetries = maxRetries;
    this.backoffCoefficient = backoffCoefficient;
    this.initialDelayMs = initialDelayMs;
    this.maxDelayMs = maxDelayMs;
  }
}

export function parseRetries(
  retries: number | Retries | undefined,
): Retries | undefined {
  if (retries === undefined) return undefined;
  if (typeof retries === "number") {
    if (!Number.isInteger(retries) || retries < 0 || retries > 10) {
      throw new Error(
        `Retries parameter must be an integer between 0 and 10. Found: ${retries}`,
      );
    }
    return new Retries({
      maxRetries: retries,
      backoffCoefficient: 1.0,
      initialDelayMs: 1000,
    });
  }
  if (retries instanceof Retries) return retries;
  throw new Error(
    `Retries parameter must be an integer or instance of Retries. Found: ${typeof retries}.`,
  );
}
